controller/api/destination/watcher/ip_watcher.go

```diff
@@ -0,0 +1,372 @@
+////////////////////////
+
+// Subscribe to an authority.
+// The provided listener will be updated each time the address set for the
+// given authority is changed.
+func (iw *IPWatcher) Subscribe(clusterIP string, port Port, listener EndpointUpdateListener) error {
+	iw.log.Infof("Establishing watch on service cluster ip [%s:%d]", clusterIP, port)
+	ss := iw.getOrNewServiceSubscriptions(clusterIP)
+	return ss.subscribe(port, listener)
+}
+
+// Unsubscribe removes a listener from the subscribers list for this authority.
+func (iw *IPWatcher) Unsubscribe(clusterIP string, port Port, listener EndpointUpdateListener) {
+	iw.log.Infof("Stopping watch on service cluster ip [%s:%d]", clusterIP, port)
+	ss, ok := iw.getServiceSubscriptions(clusterIP)
+	if !ok {
+		iw.log.Errorf("Cannot unsubscribe from unknown service ip [%s:%d]", clusterIP, port)
+		return
+	}
+	ss.unsubscribe(port, listener)
+}
+
+func (iw *IPWatcher) addService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem || service.Spec.ClusterIP == "None" {
+		return
+	}
+
+	ss := iw.getOrNewServiceSubscriptions(service.Spec.ClusterIP)
+
+	ss.updateService(service)
+}
+
+func (iw *IPWatcher) deleteService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem {
+		return
+	}
+
+	ss, ok := iw.getServiceSubscriptions(service.Spec.ClusterIP)
+	if ok {
+		ss.deleteService()
+	}
+}
+
+func (iw *IPWatcher) addPod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if pod.Namespace == kubeSystem {
+		return
+	}
+	if pod.Status.PodIP == "" {
+		// Pod has not yet been assigned an IP address.
+		return
+	}
+	ss := iw.getOrNewServiceSubscriptions(pod.Status.PodIP)
+
+	ownerKind, ownerName := iw.k8sAPI.GetOwnerKindAndName(pod, true)
+	podSet := PodSet{
+		Pods: map[PodID]Address{
+			PodID{
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			}: Address{
+				IP:        pod.Status.PodIP,
+				Port:      0, // Will be set by individual subscriptions
+				Pod:       pod,
+				OwnerName: ownerName,
+				OwnerKind: ownerKind,
+			},
+		},
+		Labels: map[string]string{},
+	}
+
+	ss.updatePod(podSet)
+}
+
+func (iw *IPWatcher) deletePod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if pod.Namespace == kubeSystem {
+		return
+	}
+
+	ss, ok := iw.getServiceSubscriptions(pod.Status.PodIP)
+	if ok {
+		ss.deletePod()
+	}
+}
+
+// Returns the serviceSubscriptions for the given clusterIP if it exists.  Otherwise,
+// create a new one and return it.
+func (iw *IPWatcher) getOrNewServiceSubscriptions(clusterIP string) *serviceSubscriptions {
+	iw.Lock()
+	defer iw.Unlock()
+
+	// If the service doesn't yet exist, create a stub for it so the listener can
+	// be registered.
+	ss, ok := iw.publishers[clusterIP]
+	if !ok {
+		ss = &serviceSubscriptions{
+			clusterIP: clusterIP,
+			listeners: make(map[EndpointUpdateListener]Port),
+			endpoints: iw.endpoints,
+			log:       iw.log.WithField("clusterIP", clusterIP),
+		}
```

Comment:
It seems there's still an issue with pods with `hostNetwork: true`. If `clusterIP` corresponds to one or more than one such pods, even though they'll be ignored below, `ss` still gets set and will get returned at the end of this method.

Reply:
hmmm, what do you think the correct behavior should be here?

getOrNewServiceSubscriptions should always return a serviceSubscriptions even if the ip address doesn't correspond to any pods or services in the cluster.  (or it only corresponds to hostNetwork pods).  In either of these cases, the returned serviceSubscriptions should have `ss.pod = PodSet{}`.  

This causes `updatePod` to send a `listener.Add` with a size 0 pod set.  Perhaps we simply need a check here to skip the add if the size of the pod set is 0.

Is there another issue that I'm missing?

Reply:
Yeah, an issue is that in `addPod` a `podSet` is getting rebuilt, ignoring that `ss.pod` got skipped in `getOrNewServiceSubscriptions`, so when asking the service for a host IP I still get something back:
```
# 192.168.99.208 is a node IP, which has many pods with hostNetwork pointing to it,
# and the following call returns one of them
$ go run controller/script/destination-client/main.go -path 192.168.99.208 -method get -addr localhost:8086
INFO[0000] Add:
INFO[0000] labels: map[]
INFO[0000] - 192.168.99.208:80
INFO[0000]   - labels: map[deployment:emoji pod:emoji-64f4c9cd58-cv4p5 pod_template_hash:64f4c9cd58 serviceaccount:emoji]                                                                                                                    
INFO[0000]   - protocol hint: UNKNOWN
INFO[0000]
```

Reply:
ah I see what you mean.  I think this can be fixed the same way as https://github.com/linkerd/linkerd2/pull/3595#discussion_r350544363 by passing `ss.pod` into `ss.updatePod` instead.